### PR TITLE
V1 support for remaining sampling parameters

### PR DESCRIPTION
**Pull request by tchedaTT**
## Purpose
Implement the changes from https://github.com/tenstorrent/vllm/issues/301

Custom logits processors are not currently supported - this is a limitation of the v1 version we branched from. More changes possibly required once we rebase onto a version of v1 that supports them

## Test Plan

Added tests covering these sampling features, testing across DP/non DP correct so far.  CI pipeline to run them across all machine/model combos WIP.

### CODE REVIEW

File: tests/tt/test_v1_sampling.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+import pytest
+
+from tests.tt.utils import RequestConfig, run_concurrent_batch
+
+
+class TestV1Sampling:
+    """
+    Verify v1 sampling parameters work correctly via the OpenAI API.
+    """
```
Comment (skhorasganiTT) [inline tests/tt/test_v1_sampling.py]: Please rename the file/class to indicate the specific params that are being tested.

File: tests/tt/utils.py [hunk 2]
```diff
@@ -17,15 +17,41 @@ class RequestConfig:
...
-async def send_request(async_client, model: str, config: RequestConfig):
-    """Send a single async request with its own parameters."""
+    logprobs: Union[int, None] = None
+    min_p: float = 0.0
+    bad_words: Union[list[str], None] = None
+    logit_bias: Union[dict[int, float], None] = None
+    allowed_token_ids: Union[list[int], None] = None
+    min_tokens: int = 0
+    return_tokens_as_token_ids: bool = False
+    # Note: Custom logits_processors are NOT supported in V1.
```
Comment (skhorasganiTT) [inline tests/tt/utils.py]: in current version right?

File: tests/tt/test_v1_sampling.py (hunk 1 above)
Comment (tchedaTT) [reply tests/tt/test_v1_sampling.py]: Split the tests into test_host_only_params and test_logprobs and correspondingly named classes

File: tests/tt/test_v1_sampling.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+"""
+Tests for v1 sampling parameters.
+
+Note: Custom user-provided logits_processors are NOT supported in V1.
+V1 only supports built-in logits processors (min_p, logit_bias, min_tokens)
+which are configured via their respective sampling parameters.
+This has been added to V1 since the version we have checked out.
```
[the remark this answers is not in the document]
Comment (tchedaTT) [reply tests/tt/test_v1_sampling.py]: Added

File: tests/tt/test_v1_sampling.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        """Test logprobs parameter returns actual logprobs data.
+
+        Parametrized by batch size to verify logprobs work correctly across
+        all DP ranks. On multi-device setups (num_devices > 1), logprobs can
+        be computed on device. On single-device setups, logprobs fall back to
+        host sampling.
+
+        Note: Device-sampled logprobs only include the sampled token's logprob,
+        not top-k alternatives or ranks. Host-sampled logprobs include full
+        top-k alternatives and ranks.
```
[the remark this answers is not in the document]
Comment (tchedaTT) [reply tests/tt/test_v1_sampling.py]: Removed

File: tests/tt/test_v1_sampling.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+
+from tests.tt.utils import RequestConfig, run_concurrent_batch
+
+
+class TestV1Sampling:
+    """
+    Verify v1 sampling parameters work correctly via the OpenAI API.
+    """
+
+    @pytest.mark.parametrize("batch_fraction", [0, 0.5, 1, 1.5])
```
[the remark this answers is not in the document]
Comment (tchedaTT) [reply tests/tt/test_v1_sampling.py]: More than 1 batch, eg. to make sure that updates are handled correctly when a batch is composed of some new and some old requests

File: tests/tt/test_v1_sampling.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+        not top-k alternatives or ranks. Host-sampled logprobs include full
+        top-k alternatives and ranks.
+
+        batch_fraction: 1 = full batch, 0.5 = half batch, 0 = single request
+        num_logprobs: number of top logprobs alternatives to return
+        """
+        if batch_fraction == 0:
+            num_requests = 1
+        else:
+            num_requests = max(1, int(max_batch_size * batch_fraction))
```
[the remark this answers is not in the document]
Comment (tchedaTT) [reply tests/tt/test_v1_sampling.py]: right

File: tests/tt/utils.py (hunk 2 above)
Comment (tchedaTT) [reply tests/tt/utils.py]: yes - removed the note as it is out of place there

File: vllm/v1/worker/tt_input_batch.py [hunk 7]
```diff
@@ -303,13 +389,52 @@ def condense(self, empty_req_indices: list[int]) -> None:
...
+                self.generators[empty_index] = self.generators.pop(
+                    last_req_index)
+            else:
+                self.generators.pop(empty_index, None)
+
+            if last_req_index in self.bad_words_token_ids:
+                self.bad_words_token_ids[
+                    empty_index] = self.bad_words_token_ids.pop(last_req_index)
+            else:
+                self.bad_words_token_ids.pop(empty_index, None)
```
Comment (skhorasganiTT) [inline vllm/v1/worker/tt_input_batch.py]: are the pops on the empty_index needed?

File: vllm/v1/worker/tt_input_batch.py [hunk 8]
```diff
@@ -123,6 +128,34 @@ def __init__(
         # Sampling-related.
         self.sampling = SamplingInputBatch(max_num_reqs)
 
+        # req_index -> generator
+        # NOTE: The indices of the requests that do not have their own
+        # generator should not be included in the dictionary.
+        self.generators: dict[int, torch.Generator] = {}
```
Comment (skhorasganiTT) [inline vllm/v1/worker/tt_input_batch.py]: Are these generators used for anything? Can they be deleted?

File: vllm/v1/worker/tt_input_batch.py [hunk 9]
```diff
@@ -123,6 +128,34 @@ def __init__(
...
+                                                    device=torch.device("cpu"))
+
+        # Allowed token IDs tracking
+        self.has_allowed_token_ids: set[str] = set()
+        # NOTE: In the mask tensor, if the corresponding token is allowed,
+        # the value is False. Since we use masked_fill_ to set -inf.
+        self.allowed_token_ids_mask: Optional[torch.Tensor] = None
+
+        # req_index -> bad_words_token_ids
+        self.bad_words_token_ids: dict[int, list[list[int]]] = {}
```
Comment (skhorasganiTT) [inline vllm/v1/worker/tt_input_batch.py]: Can these attributes be moved inside SamplingInputBatch? The original intention was that that class contains all sampling info

File: vllm/v1/worker/tt_input_batch.py [hunk 10]
```diff
@@ -123,6 +128,34 @@ def __init__(
         # Sampling-related.
         self.sampling = SamplingInputBatch(max_num_reqs)
 
+        # req_index -> generator
+        # NOTE: The indices of the requests that do not have their own
+        # generator should not be included in the dictionary.
+        self.generators: dict[int, torch.Generator] = {}
+
+        # Logprobs tracking
+        self.num_logprobs: dict[str, int] = {}
```
Comment (skhorasganiTT) [inline vllm/v1/worker/tt_input_batch.py]: Can enable_log_probs in SamplingInputBatch be replaced with a num_logprobs tensor (not dict). Shouldn't need both. Also we'll want that anyways to support the on device case.

File: vllm/v1/sample/logits_processor.py [hunk 11]
```diff
@@ -218,6 +218,21 @@ def all(self) -> Iterator[LogitsProcessor]:
...
+        min_p, logit_bias, min_tokens which are handled by built-in processors).
+        """
+        for proc in self.all:
+            if isinstance(proc, MinPLogitsProcessor) and proc.min_p_count:
+                return True
+            if isinstance(proc, LogitBiasLogitsProcessor) and proc.biases:
+                return True
+            if isinstance(proc, MinTokensLogitsProcessor) and proc.min_toks:
+                return True
+        return False
```
Comment (skhorasganiTT) [inline vllm/v1/sample/logits_processor.py:234]: Did this exist upstream? With the upstream pull i'm no longer seeing this file

File: vllm/v1/worker/tt_model_runner.py [hunk 12]
```diff
@@ -77,6 +86,9 @@ class TTModelInput:
     # previous step (used by on-device sampling).
     reset_batch: bool = False
 
+    # max_num_logprobs: max logprobs value across all requests
+    max_num_logprobs: Optional[int] = None
```
Comment (skhorasganiTT) [inline vllm/v1/worker/tt_model_runner.py]: can this be an attribute of TTSamplingParams since it's sampling related?

File: vllm/v1/worker/tt_model_runner.py [hunk 13]
```diff
@@ -593,6 +613,11 @@ def _prepare_model_inputs(
             prompt_tokens=prompt_tokens,
             output_tokens=output_tokens,
             reset_batch=reset_batch,
+            # Host-only params - wrapped in lists for DP compatibility
```
Comment (skhorasganiTT) [inline vllm/v1/worker/tt_model_runner.py]: Host-only params -> Host-only sampling params

File: vllm/v1/worker/tt_model_runner.py [hunk 14]
```diff
@@ -593,6 +613,11 @@ def _prepare_model_inputs(
...
+            allowed_token_ids_mask_list=[allowed_token_ids_mask],
+            bad_words_token_ids_list=[input_batch.bad_words_token_ids.copy()],
```
Comment (skhorasganiTT) [inline vllm/v1/worker/tt_model_runner.py]: is the copy needed?

File: vllm/v1/worker/tt_model_runner.py [hunk 15]
```diff
@@ -612,6 +637,9 @@ def build_model_input(
         if not scheduler_output.total_num_scheduled_tokens:
             return None
 
+        # Refresh logits processors with batch state changes
+        self.input_batch.refresh_logitsprocs()
```
Comment (skhorasganiTT) [inline vllm/v1/worker/tt_model_runner.py]: can this go to the end of _update_states like the gpu_model_runner?

File: vllm/v1/worker/tt_model_runner.py [hunk 16]
```diff
@@ -717,10 +745,25 @@ def build_dp_decode_gather_input(
                 "output_tokens": model_input.output_tokens,
             }
 
+        # Host-only params for host sampling
+        host_only_params = None
```
Comment (skhorasganiTT) [inline vllm/v1/worker/tt_model_runner.py]: Can you rename host_only_params to host_only_sample_params everywhere, and Host-only params to Host-only sampling params

File: vllm/v1/worker/tt_worker.py [hunk 17]
```diff
@@ -259,16 +268,25 @@ def concat_and_execute_dp(self, inputs: Union[list[Optional[TTModelInput]],
                     ],
                                           dim=0)
             sampled_token_ids_per_dp[dp_rank] = token_ids
-        return torch.stack(sampled_token_ids_per_dp)  # [world, B, 1]
```
Comment (skhorasganiTT) [inline vllm/v1/worker/tt_worker.py:262]: plz keep the comment with the shape of the token ids

File: vllm/v1/engine/core.py [hunk 18]
```diff
@@ -1360,11 +1363,46 @@ def _execute_model_dp_gather(self, scheduler_output: SchedulerOutput):
...
+                        dist.recv(host_only_tensor, src=0, group=group)
+                        gathered_host_only_params = pickle.loads(
+                            host_only_tensor.numpy().tobytes())
+
             if local_rank == 0:
                 gathered_inputs = {
                     "int_inputs": stacked_int,
                     "float_inputs": stacked_float,
                     "sampling_tokens_inputs": gathered_tokens_inputs,
+                    "host_only_params": gathered_host_only_params,
```
Comment (skhorasganiTT) [inline vllm/v1/engine/core.py]: host_only_sample_params

File: vllm/v1/worker/tt_model_runner.py [hunk 19]
```diff
@@ -1180,13 +1309,16 @@ def execute_with_model_input(
...
-            # tt_out is a tuple of (logits, logprobs)
-            # v1 currently doesn't handle logprobs from TT models
-            if isinstance(tt_out, tuple):
-                tt_out = tt_out[0]
+
+        # tt_out can be a tuple of (logits_or_tokens, logprobs) when device
+        # sampling is enabled with logprobs. Extract both components.
+        tt_log_probs = None
+        if isinstance(tt_out, tuple):
+            tt_out, tt_log_probs = tt_out
```
Comment (skhorasganiTT) [inline vllm/v1/worker/tt_model_runner.py:1363]: should we have an assert that there are 2 outputs when perform_device_sampling and model_input.max_num_logprobs?

File: vllm/v1/worker/tt_input_batch.py (hunk 7 above)
Comment (tchedaTT) [reply vllm/v1/worker/tt_input_batch.py]: Removed - taken care of by remove_request already

File: vllm/v1/worker/tt_input_batch.py (hunk 8 above)
Comment (tchedaTT) [reply vllm/v1/worker/tt_input_batch.py]: Generators now properly forwarded

File: vllm/v1/worker/tt_input_batch.py (hunk 9 above)
Comment (tchedaTT) [reply vllm/v1/worker/tt_input_batch.py]: moved

File: vllm/v1/worker/tt_input_batch.py (hunk 10 above)
Comment (tchedaTT) [reply vllm/v1/worker/tt_input_batch.py]: changes samplinginputbatch to contain num_logprobs tensor - ttsamplingparams still contains booleans

File: vllm/v1/sample/logits_processor.py (hunk 11 above)
Comment (tchedaTT) [reply vllm/v1/sample/logits_processor.py:234]: yes, this was a part of upstream that I patched in

File: vllm/v1/worker/tt_model_runner.py (hunk 12 above)
Comment (tchedaTT) [reply vllm/v1/worker/tt_model_runner.py]: This is sampling related, but I think our current contract is that TTSamplingParams contains only the params supported on device, and so far device just expects a list of booleans

File: vllm/v1/worker/tt_model_runner.py (hunk 16 above)
Comment (tchedaTT) [reply vllm/v1/worker/tt_model_runner.py]: done

File: vllm/v1/worker/tt_worker.py (hunk 17 above)
Comment (tchedaTT) [reply vllm/v1/worker/tt_worker.py:262]: restored

File: vllm/v1/worker/tt_model_runner.py (hunk 14 above)
Comment (tchedaTT) [reply vllm/v1/worker/tt_model_runner.py]: no, removed

File: vllm/v1/worker/tt_model_runner.py (hunk 19 above)
Comment (tchedaTT) [reply vllm/v1/worker/tt_model_runner.py:1363]: Good idea, added

File: vllm/v1/worker/tt_model_runner.py (hunk 15 above)
Comment (tchedaTT) [reply vllm/v1/worker/tt_model_runner.py]: Yes, good catch - I think the current state could actually be a bug if our batch went full->empty->full . Moved.

**Comment by tchedaTT** [comment]
new pipeline: https://github.com/tenstorrent/tt-metal/actions/runs/21972598335

**Comment by tchedaTT** [comment]
previous run had multiple unrelated failures, new one: https://github.com/tenstorrent/tt-metal/actions/runs/21996024453

File: vllm/v1/worker/tt_model_runner.py [hunk 20]
```diff
@@ -1050,8 +1195,12 @@ def execute_model(
             return EMPTY_MODEL_RUNNER_OUTPUT
 
         # Only 1 DP rank here
-        sampled_token_ids = self.execute_with_model_input(model_input)[0]
-        output = self.generate_runner_output(sampled_token_ids)
+        sampled_token_ids_per_dp, logprobs_per_dp = self.execute_with_model_input(  # noqa: E501
+            model_input)
+        sampled_token_ids = sampled_token_ids_per_dp[0]
+        logprobs_tensors = logprobs_per_dp[0] if logprobs_per_dp else None
+        logprobs = logprobs_tensors.tolists() if logprobs_tensors else None
```
Comment (skhorasganiTT) [inline vllm/v1/worker/tt_model_runner.py:1202]: is the tolists() needed here?

**Review by skhorasganiTT — APPROVED** [review]

File: vllm/v1/worker/tt_model_runner.py (hunk 20 above)
Comment (tchedaTT) [reply vllm/v1/worker/tt_model_runner.py:1202]: This is here to match the conversion done in the DP path, which in turn is there to match what the gpu_model_runner does. It might be cleaner to do it in get_output_tokens and investigate whether it can be removed, but I think that's best handled in a separate PR.
